lib/spack/spack/spec.py

```diff
@@ -2225,7 +2225,10 @@ def validate_or_raise(self):
             if not spec.virtual:
                 pkg_cls = spec.package_class
                 pkg_variants = pkg_cls.variants
-                not_existing = set(spec.variants) - set(pkg_variants)
+                # 'patches' is a variant that may be set on any package but is
+                # not necessarily recorded by the package's class
+                not_existing = set(spec.variants) - (
+                    set(pkg_variants) | set(['patches']))
```

Comment:
can this use `spack.directives.reserved_names` so it's not a hard-coded special case?